### PR TITLE
AP_ParticleMatter: Add I2C driver for Panasonic SN-GCJA5

### PR DESCRIPTION
**Pull request by jmachuca77**
This PR add support for the [Panasonic SN-GCJA5](https://www.sparkfun.com/products/17123) Laser Type Particle Matter sensor. It adds a log message to log the readings from the sensor so that it can be used in surveys.
17123-Particle_Sensor_-_SN-GCJA5-01

This has been flight tested by Oliver Wolkman on plane and rover.

### CODE REVIEW

**Comment by WickedShell** [comment]
I don't see a path where this helps any of the core flight code. I think this is probably best done by using @IamPete1's I2C bindings for scripting and leaving the logging of the sensor as a script, rather then adding it to ArduPilot as a core library/dependency we have to maintain.

File: ArduCopter/Copter.cpp [hunk 1]
```diff
@@ -164,6 +164,7 @@ const AP_Scheduler::Task Copter::scheduler_tasks[] = {
 #if RPM_ENABLED == ENABLED
     SCHED_TASK(rpm_update,            40,    200),
 #endif
+    SCHED_TASK_CLASS(AP_PM_SNGCJA5,       &copter.sngcja5,          update,          1,  100),
```
Comment (khancyr) [inline ArduCopter/Copter.cpp:167]: this should at least be unclosed by some #if to not be included by default

File: libraries/AP_ParticleMatter/AP_PM_SNGCJA5.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+#define SNGCJA5_PC_10_0_H 0x19
+
+void AP_PM_SNGCJA5::init()
+{
+    FOREACH_I2C(i) {
+        if (init(i)) {
+            return;
+        }
+    }
+    gcs().send_text(MAV_SEVERITY_INFO, "No SNGCJA5 found");
```
Comment (khancyr) [inline libraries/AP_ParticleMatter/AP_PM_SNGCJA5.cpp:57]: this will spam

File: libraries/AP_ParticleMatter/AP_PM_SNGCJA5.cpp [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    uint8_t val[1];
+    if (!dev->read_registers(SNGCJA5_STATUS, val, sizeof(val))) {
+        return;
+    }
+
+    uint8_t ALL_REGS[24];
+    if (!dev->read_registers(SNGCJA5_PM1_0_LL, ALL_REGS, sizeof(ALL_REGS))) {
+        return;
+    }
+    double  PM1_0  = (ALL_REGS[0] | ALL_REGS[1] << 8 | ALL_REGS[2] << 16 | ALL_REGS[3] << 24);
```
Comment (khancyr) [inline libraries/AP_ParticleMatter/AP_PM_SNGCJA5.cpp:87]: pretty sure we don't need double here ?

File: libraries/AP_ParticleMatter/AP_PM_SNGCJA5.cpp [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    double  PM2_5  = (ALL_REGS[4] | ALL_REGS[5] << 8 | ALL_REGS[6] << 16 | ALL_REGS[7] << 24);
+    double  PM10_0 = (ALL_REGS[8] | ALL_REGS[9] << 8 | ALL_REGS[10] << 16 | ALL_REGS[11] << 24);
+    int     PC0_5  = (ALL_REGS[12] | ALL_REGS[13] << 8);
```
Comment (khancyr) [inline libraries/AP_ParticleMatter/AP_PM_SNGCJA5.cpp:90]: int16_t ?

File: libraries/AP_ParticleMatter/AP_PM_SNGCJA5.cpp [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+{
+    dev = std::move(hal.i2c_mgr->get_device(bus, SNGCJA5_I2C_ADDRESS));
+    if (!dev) {
+        return false;
+    }
+
+    // read at 1Hz
+    printf("Starting Particle Matter Sensor on I2C\n");
+
+    dev->register_periodic_callback(1000000, FUNCTOR_BIND_MEMBER(&AP_PM_SNGCJA5::read_frames, void));
```
Comment (khancyr) [inline libraries/AP_ParticleMatter/AP_PM_SNGCJA5.cpp:70]: so we get it from the thread and from the vehicle code at 1Hz ?

**Review by khancyr — CHANGES_REQUESTED** [review]
The SITL backend isn't write.

As @WickedShell said, this doesn't bring any value to the autopilot. I understand it can be useful on a rover to map the PM density in a house but it doesn't serve on general purpose.
Now that we have LUA script, IOT sensor should be on LUA

**Comment by tridge** [comment]
This driver should be done as a Lua script. The driver as written uses flash space on all boards, whereas if done as a lua script then it only gets used on boards where it is installed.
The general answer if we don't think drivers for sensors that have nothing to do with flight control should be in the core firmware, instead they should be done as example lua scripts.

**Comment by rmackay9** [comment]
@jmachuca77, by the way, PeterB added a PR a few months ago that added support for a air quality sensor and we didn't merge it for the same reason.  So you're not alone, but I guess for now at least, we're hoping developers will be ok with writing a lua script for sensors that aren't used directly for flight control.

**Comment by rmackay9** [comment]
I think I will close this if that's OK.  Of course there's a lot of work that's gone into this so it is a bit of a shame in that way but as mentioned above at this time we're not looking to add support for drivers that aren't used in some way for flight control.  I'll be monitoring responses of course and so happy to re-open if more discussion is required.

**Pull request closed without merging**